gapic/src/main/com/google/gapid/perfetto/views/TrackContainer.java

```diff
@@ -328,32 +328,36 @@ public Dragger onDragStart(double x, double y, int mods) {
     public Hover onMouseMove(Fonts.TextMeasurer m, double x, double y) {
       if (y < TITLE_HEIGHT && (expanded || x < LABEL_WIDTH)) {
         hovered = true;
-        double p = m.measure(Fonts.Style.Normal, summary.getTitle()).w + LABEL_OFFSET;
+        double textX =
+            Math.ceil(m.measure(Fonts.Style.Normal, summary.getTitle()).w) + LABEL_OFFSET;
+        double toggleX = (expanded && filter != null) ?
+            Math.max(LABEL_TOGGLE_X, textX) + LABEL_ICON_SIZE : textX;
+        double pinX = Math.max(LABEL_PIN_X, toggleX) + LABEL_ICON_SIZE;
+        double redraw = (pinX > LABEL_WIDTH) ? pinX + LABEL_MARGIN : 0;
         if (expanded) {
-          if (x < p) {
-            return new TrackTitleHover(Hover.NONE, () -> expanded = false);
+          if (x < textX) {
+            return new TrackTitleHover(Hover.NONE, redraw, () -> expanded = false);
           }
-          p = Math.max(p + LABEL_MARGIN, LABEL_TOGGLE_X);
         } else {
-          if (x < Math.min(p, LABEL_PIN_X - LABEL_MARGIN)) {
-            return new TrackTitleHover(summary.onMouseMove(m, x, y), () -> expanded = true);
+          if (x < Math.min(textX, LABEL_PIN_X - LABEL_MARGIN)) {
+            return new TrackTitleHover(summary.onMouseMove(m, x, y), redraw, () -> expanded = true);
           }
-          p = LABEL_PIN_X;
+          toggleX = LABEL_PIN_X;
+          pinX = LABEL_WIDTH;
         }
-        if (expanded && filter != null && x >= p && x < p + LABEL_ICON_SIZE) {
-          return new TrackTitleHover(Hover.NONE, () -> {
+        if (expanded && filter != null && x >= textX && x < toggleX) {
+          return new TrackTitleHover(Hover.NONE, redraw, () -> {
             filtered = !filtered;
             filter.accept(detail, filtered);
           });
         }
-        p = Math.max(p, LABEL_PIN_X);
-        if (x >= p && x < p + LABEL_ICON_SIZE) {
-          return new TrackTitleHover(Hover.NONE, () -> pinState.toggle(this::copy));
+        if (x >= toggleX && x < pinX) {
```

Comment:
Clicking on empty area between text and pin icon will also result in pinning action (e.g. in GPU title). Not sure if it's expected behavior, if not, maybe `(x >= pinX - LABEL_ICON_SIZE && x < pinX)` ?

Reply:
Done. Nice catch!